add Observatory via CLI using JSON file

Maybe a command line driving thing? allowing members of telescopes to add their own telescopes config data to the database the UI utilises to process telescropes.

[jodrell_bank_alt.json](https://github.com/user-attachments/files/16144128/jodrell_bank_alt.json)

Using the above file to attempt to create a new observatory via the CLI leads to the following error:

`Caused by: com.fasterxml.jackson.databind.JsonMappingException: Already had POJO for id (java.lang.Long) [[ObjectId: key=0, type=com.fasterxml.jackson.databind.deser.impl.PropertyBasedObjectIdGenerator, scope=org.ivoa.dm.proposal.prop.Telescope]] (through reference chain: org.ivoa.dm.proposal.prop.Observatory["telescopes"]->java.util.ArrayList[1]->org.ivoa.dm.proposal.prop.Telescope["_id"])
`

However, using the identical json string in the Swagger UI for the API succeeds in creating the Observatory.

This is the relevant implementation detail of the 'createObservatory' command in the CLI:

```
Observatory observatory = parent.mapper.readValue(jsonFile, Observatory.class);
System.out.println(parent.mapper.writeValueAsString(parent.api.createObservatory(observatory)));
```
I don't know how to fix the problem.

--- a/jvm/src/main/java/org/orph2020/pst/cli/CreateObservatory.java
+++ b/jvm/src/main/java/org/orph2020/pst/cli/CreateObservatory.java
@@ -30,8 +30,6 @@ public class CreateObservatory implements Runnable{
                     inputObservatory.getIvoid(), inputObservatory.getWikiId()
             );
 
-            System.out.println(parent.mapper.writerWithDefaultPrettyPrinter().writeValueAsString(observatory));
-
             Observatory persistedObservatory = parent.api.createObservatory(observatory);
             Long id = persistedObservatory.getId();
 
@@ -55,10 +53,10 @@ public class CreateObservatory implements Runnable{
                 parent.api.createAndAddArray(id, telescopeArray);
             }
 
-            System.out.println(parent.mapper
-                    .writerWithDefaultPrettyPrinter()
-                    .writeValueAsString(parent.api.getObservatory(id))
-            );
+            Observatory createdObservatory = parent.api.getObservatory(id);
+
+            System.out.println("Observatory '" + createdObservatory.getName() + "' has been created " +
+                    "with id: " + createdObservatory.getId());
 
         } catch (Exception e) {
             throw new RuntimeException(e);

--- a/jvm/src/main/java/org/orph2020/pst/cli/CreateObservatory.java
+++ b/jvm/src/main/java/org/orph2020/pst/cli/CreateObservatory.java
@@ -24,6 +24,8 @@ public class CreateObservatory implements Runnable{
 
             Observatory observatory = parent.mapper.readValue(jsonFile, Observatory.class);
 
+            System.out.println(parent.mapper.writerWithDefaultPrettyPrinter().writeValueAsString(observatory));
+
             System.out.println(parent.mapper.writeValueAsString(parent.api.createObservatory(observatory)));
 
         } catch (Exception e) {

--- a/jvm/src/main/java/org/orph2020/pst/cli/CreateObservatory.java
+++ b/jvm/src/main/java/org/orph2020/pst/cli/CreateObservatory.java
@@ -1,0 +1,33 @@
+package org.orph2020.pst.cli;
+
+import org.ivoa.dm.proposal.prop.Observatory;
+import picocli.CommandLine.*;
+
+import java.io.File;
+
+@Command(name = "createObservatory",
+        description = "create an Observatory in the Polaris tool from a JSON file")
+public class CreateObservatory implements Runnable{
+
+    @ParentCommand
+    private PolarisCLI parent;
+
+    @Option(names = {"-f", "--jsonFile"}, required = true, paramLabel = "JSONFILE",
+            description = "the file containing the Observatory defined in JSON")
+    File jsonFile;
+
+    @Override
+    public void run() {
+        try {
+            System.out.println("Creating Observatory");
+            System.out.println("Observatory file input: " + jsonFile.getAbsolutePath());
+
+            Observatory observatory = parent.mapper.readValue(jsonFile, Observatory.class);
+
+            System.out.println(parent.mapper.writeValueAsString(parent.api.createObservatory(observatory)));
+
+        } catch (Exception e) {
+            throw new RuntimeException(e);
+        }
+    }
+}

--- a/jvm/src/main/java/org/orph2020/pst/cli/CreateObservatory.java
+++ b/jvm/src/main/java/org/orph2020/pst/cli/CreateObservatory.java
@@ -1,9 +1,10 @@
 package org.orph2020.pst.cli;
 
-import org.ivoa.dm.proposal.prop.Observatory;
+import org.ivoa.dm.proposal.prop.*;
 import picocli.CommandLine.*;
 
 import java.io.File;
+import java.util.List;
 
 @Command(name = "createObservatory",
         description = "create an Observatory in the Polaris tool from a JSON file")
@@ -22,11 +23,42 @@ public class CreateObservatory implements Runnable{
             System.out.println("Creating Observatory");
             System.out.println("Observatory file input: " + jsonFile.getAbsolutePath());
 
-            Observatory observatory = parent.mapper.readValue(jsonFile, Observatory.class);
+            Observatory inputObservatory = parent.mapper.readValue(jsonFile, Observatory.class);
+
+            Observatory observatory = new Observatory(null, null, null, null,
+                    inputObservatory.getName(), inputObservatory.getAddress(),
+                    inputObservatory.getIvoid(), inputObservatory.getWikiId()
+            );
 
             System.out.println(parent.mapper.writerWithDefaultPrettyPrinter().writeValueAsString(observatory));
 
-            System.out.println(parent.mapper.writeValueAsString(parent.api.createObservatory(observatory)));
+            Observatory persistedObservatory = parent.api.createObservatory(observatory);
+            Long id = persistedObservatory.getId();
+
+            List<Telescope> telescopes = inputObservatory.getTelescopes();
+            for (Telescope telescope : telescopes) {
+                parent.api.createAndAddTelescopeToObservatory(id, telescope);
+            }
+
+            List<Instrument> instruments = inputObservatory.getInstruments();
+            for (Instrument instrument : instruments) {
+                parent.api.createAndAddInstrumentToObservatory(id, instrument);
+            }
+
+            List<Backend> backends = inputObservatory.getBackends();
+            for (Backend backend : backends) {
+                parent.api.createAndAddBackend(id, backend);
+            }
+
+            List<TelescopeArray> telescopeArrays = inputObservatory.getArrays();
+            for (TelescopeArray telescopeArray : telescopeArrays) {
+                parent.api.createAndAddArray(id, telescopeArray);
+            }
+
+            System.out.println(parent.mapper
+                    .writerWithDefaultPrettyPrinter()
+                    .writeValueAsString(parent.api.getObservatory(id))
+            );
 
         } catch (Exception e) {
             throw new RuntimeException(e);

--- a/jvm/src/main/java/org/orph2020/pst/cli/FetchObservatory.java
+++ b/jvm/src/main/java/org/orph2020/pst/cli/FetchObservatory.java
@@ -1,0 +1,23 @@
+package org.orph2020.pst.cli;
+
+import com.fasterxml.jackson.core.JsonProcessingException;
+import picocli.CommandLine.*;
+
+@Command(name="fetchObservatory")
+public class FetchObservatory implements Runnable {
+
+    @ParentCommand
+    private PolarisCLI parent;
+
+    @Override
+    public void run() {
+        try {
+            System.out.println("Fetching observatories");
+            System.out.println(parent.mapper.writeValueAsString(parent.api.getObservatories()));
+            System.out.println("Fetching observatory 3:");
+            System.out.println(parent.mapper.writeValueAsString(parent.api.getObservatory(3L)));
+        } catch (JsonProcessingException e) {
+            throw new RuntimeException(e);
+        }
+    }
+}

--- a/jvm/src/main/java/org/orph2020/pst/cli/FetchPeople.java
+++ b/jvm/src/main/java/org/orph2020/pst/cli/FetchPeople.java
@@ -1,29 +1,20 @@
 package org.orph2020.pst.cli;
 
 import com.fasterxml.jackson.core.JsonProcessingException;
-import com.fasterxml.jackson.databind.ObjectMapper;
-import jakarta.enterprise.context.Dependent;
-import jakarta.inject.Inject;
-import org.eclipse.microprofile.rest.client.inject.RestClient;
-import org.orph2020.pst.apiimpl.client.ProposalRestAPI;
-import picocli.CommandLine;
+import picocli.CommandLine.*;
 
-@Dependent
-@CommandLine.Command(name="fetchPeople")
+@Command(name="fetchPeople")
 public class FetchPeople implements Runnable{
 
-    @Inject
-    protected ObjectMapper mapper;
-
-    @RestClient
-    ProposalRestAPI api;
+    @ParentCommand
+    private PolarisCLI parent;
 
 
     @Override
     public void run() {
         try {
             System.out.println("Fetching people:");
-            System.out.println(mapper.writeValueAsString(api.getPeople()));
+            System.out.println(parent.mapper.writeValueAsString(parent.api.getPeople()));
         } catch (JsonProcessingException e) {
             throw new RuntimeException(e);
         }

--- a/jvm/src/main/java/org/orph2020/pst/cli/FetchProposal.java
+++ b/jvm/src/main/java/org/orph2020/pst/cli/FetchProposal.java
@@ -4,31 +4,19 @@ package org.orph2020.pst.cli;
  */
 
 import com.fasterxml.jackson.core.JsonProcessingException;
-import com.fasterxml.jackson.databind.ObjectMapper;
-import jakarta.enterprise.context.Dependent;
-import org.eclipse.microprofile.rest.client.inject.RestClient;
-import org.orph2020.pst.apiimpl.client.ProposalRestAPI;
 import picocli.CommandLine.*;
 
-import jakarta.inject.Inject;
-
-@Dependent
 @Command(name = "fetchProposal")
 public class FetchProposal implements Runnable {
 
-
-    @Inject
-    protected ObjectMapper mapper;
-
-    @RestClient
-    ProposalRestAPI  apiService;
-
+    @ParentCommand
+    private PolarisCLI parent;
 
     @Override
     public void run() {
        try {
-          System.out.println(mapper.writeValueAsString(apiService.getObservatories()));
-          System.out.println(mapper.writeValueAsString(apiService.getObservingProposal(1)));
+           System.out.println("Fetching proposal");
+           System.out.println(parent.mapper.writeValueAsString(parent.api.getObservingProposal(1)));
        } catch (JsonProcessingException e) {
           throw new RuntimeException(e);
        }

--- a/jvm/src/main/java/org/orph2020/pst/cli/PolarisCLI.java
+++ b/jvm/src/main/java/org/orph2020/pst/cli/PolarisCLI.java
@@ -12,11 +12,14 @@ import picocli.CommandLine.*;
 
 @QuarkusMain
 @Command(
+        name = "polaris-cli",
         mixinStandardHelpOptions = true,
         subcommands = {
+                HelpCommand.class,
                 FetchPeople.class,
                 FetchProposal.class,
-                FetchObservatory.class
+                FetchObservatory.class,
+                CreateObservatory.class
         })
 public class PolarisCLI implements QuarkusApplication, Runnable {
 

--- a/jvm/src/main/java/org/orph2020/pst/cli/PolarisCLI.java
+++ b/jvm/src/main/java/org/orph2020/pst/cli/PolarisCLI.java
@@ -1,31 +1,39 @@
 package org.orph2020.pst.cli;
 
 
+import com.fasterxml.jackson.databind.ObjectMapper;
 import io.quarkus.runtime.QuarkusApplication;
 import io.quarkus.runtime.annotations.QuarkusMain;
 import jakarta.inject.Inject;
+import org.eclipse.microprofile.rest.client.inject.RestClient;
+import org.orph2020.pst.apiimpl.client.ProposalRestAPI;
 import picocli.CommandLine;
+import picocli.CommandLine.*;
 
 @QuarkusMain
-@CommandLine.Command( mixinStandardHelpOptions = true,
-        subcommands = {FetchPeople.class, FetchProposal.class})
+@Command(
+        mixinStandardHelpOptions = true,
+        subcommands = {
+                FetchPeople.class,
+                FetchProposal.class,
+                FetchObservatory.class
+        })
 public class PolarisCLI implements QuarkusApplication {
 
-    @CommandLine.Option(names = {"-u","--user"})
+    @Option(names = {"-u","--user"})
     String user;
 
-    @CommandLine.Option(names = {"-p","--password"}, arity = "0..1", interactive = true)
+    @Option(names = {"-p","--password"}, arity = "0..1", interactive = true)
     String password;
 
     @Inject
-    CommandLine.IFactory factory;
+    IFactory factory;
 
     @Inject
-    FetchPeople fetchPeople;
+    ObjectMapper mapper;
 
-    @Inject
-    FetchProposal fetchProposal;
-
+    @RestClient
+    ProposalRestAPI api;
 
     @Override
     public int run(String... args) throws Exception {

--- a/jvm/src/main/java/org/orph2020/pst/cli/PolarisCLI.java
+++ b/jvm/src/main/java/org/orph2020/pst/cli/PolarisCLI.java
@@ -18,7 +18,7 @@ import picocli.CommandLine.*;
                 FetchProposal.class,
                 FetchObservatory.class
         })
-public class PolarisCLI implements QuarkusApplication {
+public class PolarisCLI implements QuarkusApplication, Runnable {
 
     @Option(names = {"-u","--user"})
     String user;
@@ -39,5 +39,12 @@ public class PolarisCLI implements QuarkusApplication {
     public int run(String... args) throws Exception {
         return new CommandLine(this, factory)
                 .execute(args);
+    }
+
+    @Override
+    public void run() {
+        System.out.println("Running Polaris CLI");
+        System.out.println("User: " + user);
+        System.out.println("Password: " + password);
     }
 }

--- a/jvm/src/main/resources/application.properties
+++ b/jvm/src/main/resources/application.properties
@@ -1,6 +1,9 @@
 #quarkus.application.name=pst-cli
 %dev.quarkus.rest-client.proposal-api.url=http://localhost:8084/pst/api/
 %dev.quarkus.rest-client.proposal-api.verify-host=false
+
+%test.quarkus.rest-client.proposal-api.url=http://localhost:8084/pst/api/
+
 quarkus.rest-client.proposal-api.scope=javax.inject.Singleton
 #view of the api from outside world
 quarkus.rest-client.proposal-api.url="https://kilburn.jb.man.ac.uk/pst/api/

--- a/jvm/src/test/data/jodrell_bank_alt.json
+++ b/jvm/src/test/data/jodrell_bank_alt.json
@@ -1,0 +1,300 @@
+{
+  "@type": "proposal:Observatory",
+  "name": "Jodrell Bank Alt",
+  "address": "Cheshire, UK, Earth",
+  "ivoid": {
+    "value": "ivo://obs/anobs"
+  },
+  "telescopes": [
+    {
+      "name": "Lovell",
+      "location": {
+        "@type": "coords:RealCartesianPoint",
+        "x": {
+          "@type": "ivoa:RealQuantity",
+          "unit": {
+            "value": "m"
+          },
+          "value": 3822252.643
+        },
+        "y": {
+          "@type": "ivoa:RealQuantity",
+          "unit": {
+            "value": "m"
+          },
+          "value": -153995.683
+        },
+        "z": {
+          "@type": "ivoa:RealQuantity",
+          "unit": {
+            "value": "m"
+          },
+          "value": 5086051.443
+        }
+      }
+    },
+    {
+      "name": "Mk2",
+      "location": {
+        "@type": "coords:RealCartesianPoint",
+        "x": {
+          "@type": "ivoa:RealQuantity",
+          "unit": {
+            "value": "m"
+          },
+          "value": 3822473.365
+        },
+        "y": {
+          "@type": "ivoa:RealQuantity",
+          "unit": {
+            "value": "m"
+          },
+          "value": -153692.318
+        },
+        "z": {
+          "@type": "ivoa:RealQuantity",
+          "unit": {
+            "value": "m"
+          },
+          "value": 5085851.303
+        }
+      }
+    },
+    {
+      "name": "Knockin",
+      "location": {
+        "@type": "coords:RealCartesianPoint",
+        "x": {
+          "@type": "ivoa:RealQuantity",
+          "unit": {
+            "value": "m"
+          },
+          "value": 3859711.503
+        },
+        "y": {
+          "@type": "ivoa:RealQuantity",
+          "unit": {
+            "value": "m"
+          },
+          "value": -201995.077
+        },
+        "z": {
+          "@type": "ivoa:RealQuantity",
+          "unit": {
+            "value": "m"
+          },
+          "value": 5056134.251
+        }
+      }
+    },
+    {
+      "name": "Defford",
+      "location": {
+        "@type": "coords:RealCartesianPoint",
+        "x": {
+          "@type": "ivoa:RealQuantity",
+          "unit": {
+            "value": "m"
+          },
+          "value": 3923069.171
+        },
+        "y": {
+          "@type": "ivoa:RealQuantity",
+          "unit": {
+            "value": "m"
+          },
+          "value": -146804.368
+        },
+        "z": {
+          "@type": "ivoa:RealQuantity",
+          "unit": {
+            "value": "m"
+          },
+          "value": 5009320.528
+        }
+      }
+    },
+    {
+      "name": "Pickmere",
+      "location": {
+        "@type": "coords:RealCartesianPoint",
+        "x": {
+          "@type": "ivoa:RealQuantity",
+          "unit": {
+            "value": "m"
+          },
+          "value": 3817176.561
+        },
+        "y": {
+          "@type": "ivoa:RealQuantity",
+          "unit": {
+            "value": "m"
+          },
+          "value": -162921.179
+        },
+        "z": {
+          "@type": "ivoa:RealQuantity",
+          "unit": {
+            "value": "m"
+          },
+          "value": 5089462.057
+        }
+      }
+    },
+    {
+      "name": "Darnhall",
+      "location": {
+        "@type": "coords:RealCartesianPoint",
+        "x": {
+          "@type": "ivoa:RealQuantity",
+          "unit": {
+            "value": "m"
+          },
+          "value": 3828714.513
+        },
+        "y": {
+          "@type": "ivoa:RealQuantity",
+          "unit": {
+            "value": "m"
+          },
+          "value": -169458.995
+        },
+        "z": {
+          "@type": "ivoa:RealQuantity",
+          "unit": {
+            "value": "m"
+          },
+          "value": 5080647.749
+        }
+      }
+    },
+    {
+      "name": "Cambridge",
+      "location": {
+        "@type": "coords:RealCartesianPoint",
+        "x": {
+          "@type": "ivoa:RealQuantity",
+          "unit": {
+            "value": "m"
+          },
+          "value": 3919982.752
+        },
+        "y": {
+          "@type": "ivoa:RealQuantity",
+          "unit": {
+            "value": "m"
+          },
+          "value": 2651.982
+        },
+        "z": {
+          "@type": "ivoa:RealQuantity",
+          "unit": {
+            "value": "m"
+          },
+          "value": 5013849.826
+        }
+      }
+    }
+  ],
+  "instruments": [
+    {
+      "name": "L-Band Receiver",
+      "kind": "continuum",
+      "frequencyCoverage": {
+        "start": {
+          "@type": "ivoa:RealQuantity",
+          "unit": {
+            "value": "GHz"
+          },
+          "value": 1.2
+        },
+        "end": {
+          "@type": "ivoa:RealQuantity",
+          "unit": {
+            "value": "GHz"
+          },
+          "value": 1.7
+        },
+        "spectralResolution": {
+          "@type": "ivoa:RealQuantity",
+          "unit": {
+            "value": "GHz"
+          },
+          "value": 0.5
+        },
+        "isSkyFrequency": true,
+        "polarization": "V"
+      }
+    },
+    {
+      "name": "C-Band Receiver",
+      "kind": "continuum",
+      "frequencyCoverage": {
+        "start": {
+          "@type": "ivoa:RealQuantity",
+          "unit": {
+            "value": "GHz"
+          },
+          "value": 4
+        },
+        "end": {
+          "@type": "ivoa:RealQuantity",
+          "unit": {
+            "value": "GHz"
+          },
+          "value": 7
+        },
+        "spectralResolution": {
+          "@type": "ivoa:RealQuantity",
+          "unit": {
+            "value": "GHz"
+          },
+          "value": 3
+        },
+        "isSkyFrequency": true,
+        "polarization": "V"
+      }
+    },
+    {
+      "name": "K-Band Receiver",
+      "kind": "continuum",
+      "frequencyCoverage": {
+        "start": {
+          "@type": "ivoa:RealQuantity",
+          "unit": {
+            "value": "GHz"
+          },
+          "value": 20
+        },
+        "end": {
+          "@type": "ivoa:RealQuantity",
+          "unit": {
+            "value": "GHz"
+          },
+          "value": 24
+        },
+        "spectralResolution": {
+          "@type": "ivoa:RealQuantity",
+          "unit": {
+            "value": "GHz"
+          },
+          "value": 4
+        },
+        "isSkyFrequency": true,
+        "polarization": "V"
+      }
+    }
+  ],
+  "backends": [
+    {
+      "name": "Widar Correlator",
+      "parallel": true
+    }
+  ],
+  "arrays": [
+    {
+      "name": "e-MERLIN",
+      "arrayMembers": []
+    }
+  ]
+}

--- a/jvm/src/test/java/org/orph2020/pst/cli/PolarisCliTest.java
+++ b/jvm/src/test/java/org/orph2020/pst/cli/PolarisCliTest.java
@@ -1,0 +1,26 @@
+package org.orph2020.pst.cli;
+
+import io.quarkus.test.junit.main.Launch;
+import io.quarkus.test.junit.main.LaunchResult;
+import io.quarkus.test.junit.main.QuarkusMainTest;
+import org.junit.jupiter.api.Assertions;
+import org.junit.jupiter.api.Test;
+
+@QuarkusMainTest
+public class PolarisCliTest {
+
+    @Test
+    @Launch({"--user=Batman", "--password=Batcave"})
+    public void testLaunchCommand(LaunchResult result) {
+        Assertions.assertTrue(result.getOutput()
+                .contains("User: Batman\nPassword: Batcave"));
+    }
+
+    @Test
+    @Launch({"createObservatory", "--jsonFile=src/test/data/jodrell_bank_alt.json"})
+    public void testCreateObservatoryCommand(LaunchResult result) {
+        Assertions.assertTrue(result.getOutput()
+                .contains("Observatory 'Jodrell Bank Alt' has been created"));
+    }
+
+}

